\section{Digital Evolution}

% ---- What is digital evolution? ----
% - define
% - what types of hypotheses can we test?
% @AML: this paragraph could flow a little better. It's also long. Streamline + break up.
Digital evolution experiments have emerged as a powerful research framework from which evolution can be studied.
In digital evolution, self-replicating computer programs (digital organisms) compete for resources, mutate, and evolve in a computational environment \citep{wilke_biology_2002}.
Conventionally and in my work, a digital organism comprises a linear sequence of program instructions (its genome) and a set of virtual hardware components used to interpret and express those instructions. 
To reproduce, a digital organism must execute instructions that allow it to copy its genome instruction-by-instruction and then divide.
However, self-replication is imperfect, and can result in mutations in an offspring's genome.
The combination of heritable variation due to imperfect self-replication and competition for limited resources (e.g., space, CPU time, etc.) results in evolution by natural selection.

Digital organisms live, interact, and evolve in entirely artificial environments constructed by the experimenters.
One potential drawback to digital evolution approaches is that conclusions drawn from digital evolution experiments can be artifacts of their artificial environment ([note that this is also true to some extent in experimental evolution with biological organisms]) [cite - wilke and adami, 2002].
As such, experiments with digital organisms are limited to testing general principles about evolution.
Digital evolution studies cannot reveal the history of life on Earth or [??] the biology of particular biochemical organism.
However, these drawbacks are also digital evolution's strength as a research framework.
Because we have full control over digital environments, we can perform manipulations that would be challenging or even impossible to perform in biological systems.
Additionally, we can better test the generality of evolutionary hypotheses by reproducing results across biological and digital systems, disentangling general principles from effects  specific to a particular model organism [wilke and admi 2002].
The remainder of this section provides historical context for digital evolution research (as is relevant to this dissertation), an outline of the [benefits/scientific merits/??] of experimental digital evolution, and finally, highlights prior digital evolution research on phenotypic plasticity.

% --------- HISTORICAL CONTEXT ---------
% @AML: should this be a heading? if so, what is the most appropriate name?
\subsection{Historical context}

% @AML: how to do quotes properly...?
\begin{displayquote}

% @AML: good way to format this quote?
\textit{Two computer programs in their native habitat---the memory chips of a digital computer---stalk each other from address to address.}  \citep{dewdney_core_wars_1984}
% \begin{flushright}
% A.K. Dewdney describing Core War 
% \end{flushright}
\end{displayquote}

% --- Core War ---
Modern digital evolution systems can be traced back to the 1984 computer game ``Core War'' \citep{dewdney_core_wars_1984}.
In Core War, human competitors use a simplified assembly language (called Redcode) to write ``gladiatorial'' computer programs that compete for space in the simulated core memory of a computer.
To win a bout of Core War, a program needed to shut down all of the processes associated with all of its competitor programs.
Efficient self-replication proved to be a successful strategy.
Such replicator programs could quickly spawn copies of themselves [that could, in turn, replicate themselves replicated], and if one copy were destroyed, other copies would still persist. 
Replicators grew exponentially in memory, defeating other programs by eventually copying over all of them and taking over core memory.
Despite the presence of populations of self-replicator programs and competition for space, evolution did not take place in Core War because daughter programs were always exact copies of their parent (i.e., [no mutations/heritable variation]).

% --- Core World ---
% @AML: a little clunky, needs streamlining
Inspired by Core War, Rasmussen \textit{et al.} created Core World, which was similar to Core War (using the same Redcode language for representing programs and provided the same computational environment) but introduced the possibility for random mutations when a program copied itself \citep{rasmussen_core_1989,rasmussen_coreworld_1990}.
That is, the command used by replicator programs to copy themselves was imperfect, sometimes writing a random instruction instead of copying the intended instruction. 
The Core World system had the potential for evolution, and indeed, programs seemed to evolve at first.
However, the initial Core World system proved to be ill-suited for studying evolution.
Programs were brittle: almost any mutation [deleterious/harmful/broke functionality].
Such brittleness in combination with the ability of programs to copy broken code over each other frequently caused populations to go extinct.
% Even so, ...

% --- Tierra: innovations relative to Core War/Core World, major results ---
Thomas Ray's Tierra system \citep{ray_approach_1991} innovated on the design of Core World and facilitated some of the first successful evolution experiments with self-replicating computer programs.
% @AML: awkward
% Inspired by the robustness of biological genomes, 
Ray designed the programming language used in Tierra to encode the genomes of evolving programs to be more robust than the Redcode language from Core War and Core World.
As such, genomes in Tierra are more evolvable than those in Core World because mutations in daughter programs are less often lethal.
% @aml: awkward
Tierra also protects ``living'' programs from being overwritten by their competitors, requiring digital organisms to explicitly request a protected block of memory in which they could create a copy of themselves. 
Digital organisms died when the environment ran out space, at which point the oldest programs are removed from the population to make room for new programs to be born. 

% --- early studies in Tierra ---
% [Observational studies, read like observational field studies].
In initial studies with Tierra \citep{ray_approach_1991}, Ray founded populations with an ancestral program capable only of self-replication.
Competition for space dominated these early studies, and as such there was strong selection pressure for organisms to increase their replication rate. 
Early on, Ray observed organisms with shorter genomes evolved and outcompeted organisms with [longer/ancestral] genomes because [shorter genomes contained less genetic to copy, making them more efficient self-replicators].
Ray unexpectedly observed the evolution obligate parasitism where some programs co-opted the copy machinery of their competitors to copy themselves\footnote{While Ray labeled these programs as parasitic, but they are more accurately describe as cheaters because they did not directly harm the programs whose replication machinery they co-opted.}. 
An evolutionary arms race ensued.
Would-be ``host'' programs evolved mechanisms for resisting parasites, and parasites evolved to penetrate these defensive mechanisms.
[Results initially surprising because of the seeming simplicity of Tierra's setup.]

% Eventually, obligate commensal parasites evolve, which are not capable of self-replication in isolated culture, but which can replicate when cultured with normal sel-replicating creatures
% These parasites execute some parts of the code of their hosts, but cause them no direct harm, except as competitors.
% Some potential hosts have evolved immunity to the parasites, and some parasites have evolved to circumvent this immunity.
% Facultative hyper-parasites have evolved, which can self-replicate in isolated culture, but when subjective to parasitism, subvert the parasite's energy metabolism to augment their own reproduction.
% Observed evolution of creatures that can only replicate when they occur in aggregations (evolution under circumstances of high genetic relatedness)
% Social aggregations are invaded by cheaters
% Evolutionary dynamics
% Hosts and parasites cultured together demonstrate Lotka-Volterra population cycling

% --- avida built on tierra ---
The Avida Digital Evolution Platform expanded on the ideas behind Tierra but added the ability to configure complex environments and sophisticated data tracking tools  \citep{adami_evolutionary_1994,ofria_avida_2004,ofria_avida:_2009}.
% - overview of avida -
In Avida, digital organisms compete for space on a lattice of cells \citep{ofria_avida:_2009}.
When an organism reproduces, its offspring is placed in a nearby cell (for spatially structured populations) or in a random cell (for well-mixed populations), replacing any previous occupant of that cell.
As such, improvements to the speed of self-replication are advantageous in this competition for space on the lattice.
% @AML: double check that this!
Like Tierra and Core World, organisms in Avida can improve their replication rates by improving genome efficiency (e.g., using a more compact encoding).
Avida, however, introduced the concept of resources that can be ``metabolized'' by a digital organism to accelerate the rate at which it is able to express its genome (i.e., its ``metabolic rate'').
Resources in Avida are typically associated with completing designated Boolean logic computations on inputs from the environment, but have also been associated with other tasks such as [...].
Avida gives experimenters fine-grained control over how resources are configured, including their abundance \citep{cooper_evolution_2002}, spatial distribution \citep{dolson_spatial_2017}, and their metabolic effects \citep{canino-koning_evolution_2016,canino-koning_fluctuating_2019}.

The Avida system is perhaps the most popular digital evolution system and is often credited with advancing digital evolution as model system for conducting scientifically rigorous evolution experiments.
Experimental evolution studies using Avida as a model system have been well received, and topics such as the evolution of complexity \citep{adami_evolution_2000,lenski_evolutionary_2003}, sexual recombination \citep{misevic_experiments_2010}, modularity \citep{misevic_sexual_2006}, robustness \citep{lenski_genome_1999,elena_effects_2007}, and division of labor \citep{goldsby_task-switching_2012,goldsby_evolutionary_2014} have been published in top evolutionary biology venues.
[In Chapters [X] and [Y] of this dissertation, I use Avida to investigate the evolution of phenotypic plasticity.]

% @AML: don't love this subheading
% Benefits of experimental digital evolution
\subsection{Why conduct digital evolution experiments?}

Evolution experiments using digital organisms balance the speed and transparency of mathematical and computational simulations with the open-ended realism of laboratory experiments. 
[something about how digital evolution systems are individual-based and why that is useful?]
% Individual-based => things that can't be done in numerical simulations [cite - Chapter 1 Dolson thesis].
% @AML: don't love this sentence...
Here, I overview four properties of digital evolution systems that I argue make them valuable models for studying evolutionary processes, providing exemplary examples of each:

\subsubsection{Generality}
% -- Generality --

% --- What do we mean by generality and why is it useful? --- 
Digital evolution systems offer researchers the unique opportunity to study evolution in organisms that share no ancestry with carbon-based life \citep{wilke_biology_2002}.
As biologist John Maynard Smith made the case, ``So far, we have been able to study only one evolving system and we cannot wait for interstellar flight to provide us with a second. If we want to discover generalizations about evolving systems, we will have to look to artificial ones'' \citep{maynard_smith_byte-sized_1992}.
Indeed, studies of carbon-based lifeforms that all share common ancestry dominate evolutionary biology, providing limited lens with which study evolutionary processes.
By testing hypotheses across biological and digital model systems, we can disentangle general principles from the effects of specific model organisms.  

% --- Examples ---
[Examples?]
% @AML: example of study replicated across digital organisms and biological organisms?


\subsubsection{Transparency}

% --- EXEMPLARS ---
% - Data tracking
%   - Dolson - evolutionary hot spots
% - Organism interpretability (i.e., can understand programs)

% --- Benefits of data tracking ---
Digital evolution systems allow for perfect, non-invasive data tracking.
Experimenters can save complete details of evolving populations for further analysis, including mutations, genotypes, environment information, phenotypes, organism-organism interactions, organism-environment interactions, \textit{et cetera}.
By tracking parent-offspring relationships, we can analyze complete evolutionary histories within an experiment, which circumvents the historical problem of drawing evolutionary inferences using incomplete records (from frozen samples or even fossils) and extant genetic sequences.

% --- Lineages + Phylogenies ---
Many digital evolution studies inspect the complete lineages of evolved digital organisms to tease apart the mutation-by-mutation evolution of complex features \citep{lenski_evolutionary_2003,dolson_spatial_2017,grabowski_case_2013,goldsby_evolutionary_2014,pontes_evolutionary_2019}.
In an exemplary analytical undertaking, Dolson and Ofria identified hotspots of evolutionary potential (i.e., regions of an environment where novel traits disproportionately evolved) in a spatially heterogeneous environment, finding that the \textit{paths} traversed by lineages through space might explain the existence of evolutionary hotspots \citep{dolson_hotspots_2017}. 
Recently, Dolson \textit{et al.} reviewed a breadth of ancestry-based metrics and analyses that operate on lineages and phylogenies in an effort to improve our capacity to quantitatively explore evolutionary histories in digital evolution experiments \citep{dolson_interpreting_2020}.

% -- networks --
By tracking phenotypes over time, Cooper and Ofria were able to observe the real-time evolution of stable ecosystems of digital organisms under resource-limited conditions \citep{cooper_evolution_2002}.
In a similar vein, Fortuna \textit{et al.} tracked host-parasite interactions to investigate how antagonistic coevolution shapes the structure of infection networks over time \citep{fortuna_coevolutionary_2019}.

% -- Knockout analyses --
For an individual digital organism, we can perform systematic knockout analyses to identify which instructions are responsible for producing a given phenotypic outcome.
To perform a knockout, we duplicate an organism, replacing a single instruction with an inert ``no-operation'' instruction.
We then identify any phenotypic changes by contrasting phenotype of the ``knockout'' organism and the original.
Such changes provide evidence of the role that the original instruction must have played in the genome.
For example, when an organism performs a particular computational task but loses it when an instruction is knocked out, we categorize that instruction as part of that task's genetic machinery.
By knocking out every loci in a genome, we can begin to tease apart that genome's genetic architecture, isolating regions of the genome that encode particular phenotypic traits.
For example, this sort of analysis has been applied along lineages to investigate how environmental change shapes the evolution of genetic architectures in digital organisms \citep{canino-koning_evolution_2016}.

Mutational landscaping analyses go a step further than knockout analyses, allowing experimenters to fully characterize local mutational landscapes.
That is, we can systematically apply all possible one- and two-step mutations to a given genome to generate a set of possible mutant genotypes. 
We can then express each possible mutant, characterizing its phenotype and self-replication rate relative to the original.
Such analyses have been used to quantify mutational robustness [citations] and investigate the evolution of evolvability \citep{canino-koning_fluctuating_2019}. 

\subsubsection{Control}
% -- Experimental control --

Digital evolution systems facilitate experimental manipulations that go beyond what is possible in wet-lab or field experiments.
This allows researchers to empirically test hypotheses that would otherwise be relegated to theoretical analyses.
% -- Basic configuration --
Digital evolution systems allow experimenters to vary basic parameters such as population size and mutation rate.
For example, by varying mutation rate, Wilke \textit{et al.} discovered the ``survival of the flattest'' effect where high mutation rate environments selected for genomes that with slower replication rates but that were more robust to mutation \citep{wilke_flattest_2001}. 
% For example, by controlling population size, LaBar and Adami... \citep{labar_evolution_2017}

% -- Environmental control --
% - connectivity (local vs well-mixed)
% - Dolson's work
Digital evolution experiments also allow for fine-grained control over the environment.
For example, Dolson \textit{et al.} used Avida to experimentally manipulate the spatial distribution of resource availability, finding that spatial entropy and phenotypic diversity had a strong positive correlation and that spatially heterogeneous environments exhibited increased evolutionary potential \citep{dolson_spatial_2017}.
By experimentally controlling how environments changed temporally, Nahum \textit{et al.} demonstrated that a single temporary environmental change can improve fitness landscape exploration and exploitation in evolving populations of digital organisms \citep{nahum_improved_2017}.

% -- Genetic control --
Digital evolution systems also allow experimenters to monitor and manipulate mutational effects in real-time.
Covert \textit{et al.} performed real-time reversions of all deleterious mutations as they occurred to isolate their long-term effects on evolutionary outcomes \citep{covert_experiments_2013}.
Lalejini \textit{et al.} implemented a range of slip duplication mutation operators (each designed to isolate a single effect of duplication mutations) in order to tease apart why such mutations can promote the evolution of complex traits \citep{lalejini_gene_2017}.

% -- ??? --
% @AML: becoming exhausted with doing lit review... 

\subsubsection{Scale}

Modern computers allow us to observe many generations of digital evolution at tractable time scales; thousands of generations can take mere minutes as opposed to months, years, or centuries.
For example, populations of digital organisms have been used to test theoretical predictions about the expected rate of adaption over hundreds of thousands of generations~\citep{wiser_analysis_2015,wiser_boundedness_2018}.

Further, digital evolution experiments allow researchers to enact complex experimental protocols with minimal extra effort. 
That is, unlike in wet-lab experiments, computational experiment protocols can easily be automated using modern scripting tools [cite].
% - @AML: Example?? -

With the increasing accessibility of high performance computing systems [cite], it can be trivial to evolve hundreds of replicate populations for a given experimental treatment. 
Evolution is an inherently stochastic process, so increased replication provides a clearer picture of the distribution of treatment effects. 
Further, easier to identify and study rare events.
% - @AML: Examples ??? -

Even the best high performance computing systems, however, lack the parallelism of the real world.
That is, digital evolution systems cannot yet rival bacterial systems in their ability to scale to large population sizes. 
A typical population of digital organisms comprises thousands of digital organisms; however, bacterial populations in laboratory experiments [???]. 

\subsection{Phenotypic plasticity in digital organisms}

% -- Plasticity, changing environment --
Phenotypic plasticity has been the subject of many computational evolution studies (e.g. [citations]), here I focus on previous studies using self-replicating computer programs. 
Digital evolution has been used to study varied forms of phenotypic plasticity. 
% @AML: clunky, maybe break this sentence up?
[break up: Clune \textit{et al.} investigated the selective pressures that produce adaptive phenotypic plasticity in a temporally variable environment, demonstrating that adaptively plastic digital organisms that regulate which computational tasks they perform based on the environment can evolve in Avida under the following conditions:  
a fluctuating environment where environments are differentiable by reliable cues (sensory instructions), and each environment favors different phenotypic traits (performing different computational tasks) such that no single phenotype exhibits high fitness across all environments (performing certain tasks cycled between being beneficial or harmful).]
Clune \textit{et al.} also characterized two mechanisms by which digital organisms in Avida tend to achieve phenotypic plasticity.
First, \textit{dynamic-execution-flow} plasticity uses conditional logic statements (e.g., if statements) to modify which instructions are executed based on environmental conditions. 
Second, \textit{static-execution-flow} plasticity integrates sensory information into its internal ``metabolic'' pathways such that executing the same sequence of program instructions can result in different behaviors in different environmental conditions.

% -- Groups --
Genetically homogeneous groups of individuals (e.g., cells in a multicellular organism or members of a eusocial insect colony) require phenotypic plasticity to differentiate and coordinate their differentiated behavior \citep{schlichting_origins_2003}. 
Indeed, digital evolution studies have demonstrated the \textit{de novo} evolution of adaptive phenotypic plasticity that allows  ``multicellular'' collectives of digital organisms to coordinate their behavior.
Goldsby \textit{et al.} demonstrated that direct selection for task specialization in clonal groups of digital organisms promotes the evolution of differentiation and division of labor \citep{goldsby_evolution_2010}.
In addition to directly selecting for specialization, Goldsby \textit{et al.} demonstrated that task-switching costs can promote the evolution of division of labor \citep{goldsby_evolution_2010,goldsby_task-switching_2012}.
[Underlying mechanisms by which organisms coordinated included communication, spatial patterning, and task-partitioning.]
% -- group coordination --
Digital organisms have also been used to study the evolution of synchronization and desynchronization \citep{knoester_evolution_2011} and the evolution of consensus \citep{knoester_genetic_2013} in groups of genetically homogeneous individuals; [both of which [applied algorithms applications like wireless sensor networks]. 

% -- Temporal polyethism --
Digital organisms have also been used to study the evolutionary conditions that give rise to temporal polyethism, a form of behavioral plasticity exhibited by many eusocial insect species whereby the tasks an individual attempts to perform are correlated with the individual's age.
Goldsby \textit{et al.} demonstrated that differential task-riskiness is sufficient to promote the evolution of temporal polyethism in genetically homogeneous groups \citep{goldsby_temporal_polyethism_2012}.
Individuals within a group used control-flow instructions to regulate task performance, performing low or no risk tasks early in life and then switching to performing higher risk tasks later in life. 

% -- germ-soma division of labor --
Goldsby \textit{et al.} proposed the dirty work hypothesis, which predicts that the mutagenic effects associated with metabolism can the evolution of germ--soma differentiation in multicellular organisms \citep{goldsby_evolutionary_2014}.
Goldsby \textit{et al.} tested the dirty work hypothesis using digital organisms, finding that individuals within a multicellular group used phenotypic plasticity both to differentiate between germ and soma and to efficiently divide mutagenic tasks amongst soma cells.

% -- quorum sensing --
Quorum sensing is a form of communication found in many species of bacteria that allows individuals to regulate their actions depending on the density of the surrounding population [citations].
Beckmann \textit{et al.} demonstrated the evolution of quorum sensing in digital organisms whereby individuals adaptively suppress self-replication based on their local population density \citep{beckmann_evolving_2009,beckmann_evolution_2012}.
Johnson \textit{et al} expanded this work, showing that the evolution of quorum sensing can improve the efficacy of adaptive suicidal altruism---a strategy where an altruistic organism dies to increase the fitness of other organisms---by helping altruistic individuals to regulate when to die \citep{johnson_more_2014}.  

% -- predator prey behavioral plasticity --
Using digital organisms, Wagner \textit{et al.} investigated how predator-prey coevolution influences the subsequent evolution of behavioral strategies in predator and prey species \citep{wagner_behavioral_2014}.
Wagner \textit{et al.} found increased sensor reliance and behavioral plasticity in prey coevolved with predators than in prey evolved without predators. 
Indeed, prey seemed to exapt genetic components of evolved sense-and-flee predator avoidance strategies for sense-and-retrieve foraging strategies.

% -- so much navigation --
% Right now, not including \citep{elsberry_cockroaches_2009}, but I could add it?
Grabowski \textit{et al.} tested whether digital organisms could evolve to use information about past experiences for optimal decision making \citep{grabowski2010early}. 
Specifically, an organism's reproductive success was tied to its ability to traverse a nutrient trail.
To follow a trail, organisms needed to sense and react appropriately to environmental cues that indicated how to remain on the trail.
Grabowski \textit{et al.} found that memory usage evolved only when it provided a substantial advantage; otherwise, organisms tended adopt reflexive strategies that did not require memory.
Some memory-based strategies relied on an evolved odometry mechanism wherein organisms tracked the number of steps taken and their orientation.
Building on this previous work, Grabowski \textit{et al.} used lineage analyses to disentangle the step-by-step evolution of these odometry-based strategies \citep{grabowski_case_2013}. 

Building on Grabowski \textit{et al.}'s work, Pontes \textit{et al.} used digital organisms to investigate the selective pressures promote the evolution of associative learning \citep{pontes_investigations_2017,pontes_evolutionary_2019}.
Pontes \textit{et al.} evolved organisms capable of associating novel environmental cues with their meaning in different contexts.
Environments that were stable across generations promoted the evolution of purely reflexive behavior, and environments that varied across generations (but remained stable during an organism's lifetime) promoted the evolution of learning.
Both types of environments are important for the evolution of learning, however, as Pontes \textit{et al.} found that reflexive behaviors were an important building block for the evolution of learning.  

% ------------ fin lit review ------------

The majority of work investigating phenotypic plasticity in digital organisms focuses on the particular selective pressures that can promote the evolution of adaptive phenotypic plasticity.
Some of this work also used lineage analyses to illuminate the step-by-step process by which phenotypic plasticity tended to evolve in their experiments (e.g., \citealt{grabowski_case_2013,goldsby_evolutionary_2014,pontes_evolutionary_2019}). 
Few existing studies have empirically toggled the possiblity for adaptive plasticity to evolve in order to evaluate its influence on subsequent evolutionary outcomes.
% @AML: do I need these last two sentences??
% My work presented in Chapters [x] and [y] build directly on \cite{clune_investigating_2007}.
% In [X], I 


% ? The effects of phenotypic plasticity on adaptive evolution have been disputed, as few studies have been able to observe both the initial patterns of plasticity and the subsequent divergence of traits in natural populations \citep{ghalambor_adaptive_2007,wund_assessing_2012,forsman_rethinking_2015,ghalambor_non-adaptive_2015,hendry_key_2016}.
% Experimental studies investigating the relationship between phenotypic plasticity and evolutionary outcomes can be challenging to conduct in natural systems.
% Such experiments would require the ability to irreversibly toggle plasticity followed by long periods of evolution during which detailed phenotypic data would need to be collected.
% 

